A58-client-side-weighted-round-robin-lb-policy.md

```diff
@@ -0,0 +1,251 @@
+
+In all languages, when a subchannel state is updated to `IDLE`, the control plane immediately requests a connection and treats it as if it is in `CONNECTING` for the aggregation purpose.
+
+### Subchannel Weights
+
+To enforce `blackout_period` and `weight_expiration_period`, it tracks two timestamps along with the weight. `last_updated` tracks the last time a non-zero report was received. `non_empty_since` tracks when the first non-zero load report was received, and it is reset when a subchannel comes back to `READY` or the weight expires so that it can be updated with the next non-zero load report. The blackout and expiration are enforced when the weight value is looked up. When the weight expires, `non_empty_since` is reset.
+
+The weight of a backend is calculated as ***weight = qps / cpuutilization***.
+
+```
+function UpdateWeight(qps, cpu_utilization) {
+  var new_eight =
+     cpu_utilization == 0 ? 0 : qps / cpu_utilization;
+  if (new_eight == 0) return;
+  if (non_empty_since_ == infy) non_empty_since_ = now();
+  last_updated_ = now();
+  weight_ = new_weight;
+}
+function OnSubchannelBecomesReady() {non_empty_since_ = infy;}
+function GetWeight() {
+   if (now() - last_updated_ >= weight_expiration_period_) {
+      non_empty_since_ = infy;
+      return 0;
+   }
+   if (now() - non_empty_since_ < blackout_period_) return 0;
+   return weight_;
+}
+
+```
+
+### Per-Call Load Reporting
+
+Necessary support for per-call load report handling in clients depends on [gRFC A51]. This section outlines how this features works with `weighted_round_robin`. xDS `cluster_impl` policy can serve as an example.
+
+`weighted_round_robin` defaults to per-call load reporting. The client does nothing when the per-call load report is expected but missing in the trailing metadata of the response. While `enable_oob_load_report` is set, per-call load reports are never parsed on the client side. The client does nothing when the per-call load report is present in the response when not expected.
+
+#### C/C++
+Processing per-call load reports is supported with `SubchannelCallTrackerInterface`. `weighted_round_robin` implements this interface and attaches it to a subchannel `PickResult`. Later Finish() is called back with the load metrics and updates the weight.
+
+#### Java
+`OrcaPerRequestUtil.OrcaReportingTracerFactory` is available in `io.grpc.xds.orca` to process per-call backend load reports in `ClusterImplLoadBalancer`. This needs to be moved to a new package and shared with `weighted_round_robin`. Then the picker attaches this tracer to the subchannel selection and the listener takes the reference to the subchannel to which it updates the weight to in `onLoadReport()`.
```

Comment:
There should be some way to express what needs to happen here that does not refer back to the xds_cluster_impl policy.  @ejona86

Reply:
`ClusterImplLoadBalancer` isn't involved at all today.

Reply:
Updated to only mention OrcaPerRequestUtil.OrcaReportingTracerFactory. Made similar changes to OOB and Go, PTAL.